src/middlewared/middlewared/plugins/device_/vrrp_events.py

```diff
@@ -58,28 +255,47 @@ def run(self):
                     sleep(self._retry_timeout)
 
 
-async def _start_stop_vrrp_thread(middleware):
-    global VRRP_THREAD
+async def _start_stop_vrrp_threads(middleware):
+    while not await middleware.call('system.ready'):
+        await asyncio_sleep(0.2)
 
     licensed = await middleware.call('failover.licensed')
-    if not licensed and (VRRP_THREAD is not None and VRRP_THREAD.is_alive()):
+    if not licensed:
         # maybe the system is being downgraded to non-HA
         # (this is rare but still need to handle it) or
         # system is being restarted/shutdown etc
-        await middleware.run_in_thread(VRRP_THREAD.shutdown)
-        VRRP_THREAD = None
-    elif licensed and (VRRP_THREAD is None or not VRRP_THREAD.is_alive()):
+        if VrrpObjs.fifo_thread is not None and VrrpObjs.fifo_thread.is_alive():
+            await middleware.run_in_thread(VrrpObjs.fifo_thread.shutdown)
+            VrrpObjs.fifo_thread = None
+
+        if VrrpObjs.event_thread is not None and VrrpObjs.event_thread.is_alive():
+            await middleware.run_in_thread(VrrpObjs.event_thread.shutdown)
+            VrrpObjs.event_thread = None
+
+        if VrrpObjs.event_queue is not None:
+            VrrpObjs.event_queue.clear()
+            VrrpObjs.event_queue = None
+    else:
         # if this is a system that is being licensed for HA for the
         # first time (without being rebooted) then we need to make
-        # sure we start this.
-        VRRP_THREAD = VrrpFifoThread(middleware=middleware)
-        VRRP_THREAD.start()
+        # sure we start these threads
+        if VrrpObjs.event_queue is None:
+            VrrpObjs.event_queue = deque(maxlen=1)
```

Comment:
In future consider increasing deque size so that we can keep internal history of past events to better detect flapping state.